Update doubly_linked_list.py

akashgkrishnan opened the pull request:
Update the implementation of linked list 
made changes for optimizing
added more helper methods 
used new variable names and function names

### **Describe your change:**



* [ ] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

spamegg1 commented:
>      @property
-    def is_empty(self):  
+    def is_empty(self):
         return self.head is None
Error: Process completed with exit code 1.

Add a newline at the end of the file.

akashgkrishnan commented:
can someone help me with why is it getting failed in the pre-commit pull request

akashgkrishnan commented:
@spamegg1  i have done the same but it still is failing the checks
can you confirm why
and Can you please review the PR and let me know if anything needs to be changed so i can do them as well
Thanks in advance

akashgkrishnan commented:
@spamegg1  @dhruvmanila is there something more needs to be done in this PR?

akashgkrishnan commented:
@cclauss hi sir can you review these changes?

cclauss commented on data_structures/linked_list/doubly_linked_list.py:
```diff
@@ -6,24 +6,21 @@
 
 
-class LinkedList:
-    """
```
Why remove all these tests?!?

akashgkrishnan replied:
@cclauss I had removed the previous doc test as I had completely changed the implementation. Apologies for that
I have updated the doc test with a set of new test cases and added in much more cases
Kindly review the new changes submitted for the Dubly linked List

Some functions are intentionally made of O(N) time complexity to provide a more user-friendly output on the console.
besides most functions are working in O(1) time complexity

Kindly update in case of furthermore changes are required

cclauss commented on data_structures/linked_list/doubly_linked_list.py:
```diff
@@ -33,6 +33,24 @@ def __str__(self):
+        return False
+
+    def get_head_data(self):
+        if self.head:
```
```suggestion
        return self.head.data if self.head else None
```
Same with get_tail_data()

cclauss commented on data_structures/linked_list/doubly_linked_list.py:
```diff
@@ -107,6 +125,8 @@ def delete_value(self, value):
                 self.tail = self.tail.previous
 
             self.remove_node_pointers(node)
+        else:
```
The function signature does not say that this function returns a string.  My sense is that we should raise an exception.

cclauss commented on data_structures/linked_list/doubly_linked_list.py:
```diff
@@ -33,6 +33,24 @@ def __str__(self):
             current = current.next
         return "<-->".join(str(node) for node in nodes)
```
I find the "<—>" notation to be a lot of visual clutter.  Why not just add an `.__iter__()` method and then `return str(list(self))`?

akashgkrishnan replied:
Hi sir,
I have added in a linked list iterator class for iterating over the linked list
and removed the <--> representation.

However, I have left the __str__(self) same and just removed the <--> from the string as
this would keep visualizing the linked list if someone wants to visualize the data in the Linked list while debugging as opposed to the memory address that it spits out

I had previously used the  <--> representation as I thought it would make the links between nodes more clear


@cclauss I have made the changes

cclauss commented on data_structures/linked_list/doubly_linked_list.py:
```diff
@@ -205,23 +241,12 @@ def create_linked_list():
-    >>> new_linked_list_2 = LinkedList()
-    >>> for value in range(1,10):
-    ...    new_linked_list_2.insert(value=value)
-    >>> print(new_linked_list_2)
```
Let's not lose this test!  We want a test of a linked list that is longer than one element.

akashgkrishnan commented:
@cclauss added the suggested testcase

cclauss commented:
Let's do this.  Please change the filename to `doubly_linked_list_two.py` and we can keep two different implementations of this data structure.

akashgkrishnan commented:
@cclauss renamed the filename as requested

cclauss commented:
;-) Let's not delete the current implementation.

akashgkrishnan commented:
@cclauss I'm new to this so spare me if I'm being too dumb. but I'm not sure how to rename my implementation to dll2 
and maintain the existing file as is
since I had updated the same.
cause I had updated the previous file

cclauss commented:
OK.  Can you undelete the dll.py?

cclauss commented:
If not then copy dll2 to a new pull request that refers to #2573 and then close this PR.

akashgkrishnan commented:
@cclauss new PR created at https://github.com/TheAlgorithms/Python/pull/3380
closing this PR
